### PR TITLE
Expose signature metadata to ops.

### PR DESCRIPTION
**Pull request by stellaraccident**
* Two op interfaces, one for querying instance metadata and one for getting static data needed to construct an op from a generic form.
* For torch.generic_kernel ops, metadata is splatted in during capture from Torch (it comes from the op registry, which will work for either device capture or graph import).
* Moved the 'add' out of the generated set so I can experiment on it. It implements the TorchBuildableKernelOpInterface interface which provides its metadata.
* Tomorrow I'll fill in the pass that, given these two metadata structs and a target operation name, should be able to heuristically convert any kernel call to a named op.
* Should do the right thing with mutable/immutable tensors/arrays and a few other odds and end quirks of Torch.

### CODE REVIEW

File: include/npcomp/Dialect/Torch/IR/OpInterfaces.td [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  let methods = [
+    InterfaceMethod<"Metadata for the kernel",
+      "::mlir::NPCOMP::Torch::KernelMetadata", "getTorchKernelMetadata">,
+  ];
+}
+
+def TorchBuildableKernelOpInterface :
+    OpInterface<"TorchBuildableKernelOpInterface"> {
+  let description = [{
+    Implemented by specific ops that can be constructed from a generic
```
Comment (silvasean) [inline include/npcomp/Dialect/Torch/IR/OpInterfaces.td:31]: Is it "Provides the metadata to drive the conversion." or "Provides *additional* metadata on top of TorchKernelOpInterface ..."? The comments on BuildKernelMetadata suggest the latter.

File: include/npcomp/Dialect/Torch/IR/OpInterfaces.td [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+include "mlir/IR/OpBase.td"
+
+#ifndef NPCOMP_DIALECT_ATEN_IR_TORCH_OP_INTERFACES
+#define NPCOMP_DIALECT_ATEN_IR_TORCH_OP_INTERFACES
+
+def TorchKernelOpInterface : OpInterface<"TorchKernelOpInterface"> {
+  let description = [{
+    This interface provides metadata about ops that are isomorphic with a
+    Torch kernel call. It is implemented by both the generic torch.kernel_call
```
[the remark this answers is not in the document]
Comment (stellaraccident) [reply include/npcomp/Dialect/Torch/IR/OpInterfaces.td:17]: I added quite a bit more comments that likely need to be promoted to a real document somewhere.

It is similar to the linalg.generic situation but going the other way: We start generic and seek to transform to more structured forms. Same idea but different direction/terminology.

File: include/npcomp/Dialect/Torch/IR/OpInterfaces.td (hunk 1 above)
Comment (stellaraccident) [reply include/npcomp/Dialect/Torch/IR/OpInterfaces.td:31]: Updated comments a lot. Let me know if it makes any more sense.

File: include/npcomp/Dialect/Torch/IR/OpInterfaces.h [hunk 3]
```diff
@@ -0,0 +1,81 @@
+//===- OpInterfaces.h -------------------------------------------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
```
[the remark this answers is not in the document]
Comment (stellaraccident) [reply include/npcomp/Dialect/Torch/IR/OpInterfaces.h:8]: Updated comments a lot. Let me know if it makes any more sense.

File: include/npcomp/Dialect/Torch/IR/OpInterfaces.td (hunk 1 above)
Comment (silvasean) [reply include/npcomp/Dialect/Torch/IR/OpInterfaces.td:43]: Looks a lot better, thanks!

File: include/npcomp/Dialect/Torch/IR/OpInterfaces.td (hunk 2 above)
[the remark this answers is not in the document]
Comment (silvasean) [reply include/npcomp/Dialect/Torch/IR/OpInterfaces.td:17]: LG.

File: lib/Dialect/ATen/Transforms/RecognizeKernelsPass.cpp [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+                       const BuildKernelMetadata &buildMetadata)
+        : targetOpName(targetOpName), buildMetadata(buildMetadata) {}
+    Identifier targetOpName;
+    const BuildKernelMetadata &buildMetadata;
+  };
+  using CandidateTransformList = SmallVector<CandidateTransform, 1>;
+
+  MLIRContext &context;
+  // Map of the torch.kernel_call op's kernelName() attribute to a list of
+  // candidate transforms describing how to map to a specific, recognized op.
```
Comment (silvasean) [inline lib/Dialect/ATen/Transforms/RecognizeKernelsPass.cpp:330]: maybe describe under what circumstances there can be more than one CandidateTransform?

**Review by silvasean — APPROVED** [review]
Looks good!

**Pull request merged**